src/components/DBEntryManagerPanel/AddOrEditDBEntry/index.tsx

```diff
@@ -0,0 +1,614 @@
+  entry: DBEntry,
+  // True if currently saving
+  saving: boolean,
+};
+
+/* ------------- Actions ------------ */
+
+// Types of actions
+enum ActionType {
+  // Update the DBEntry
+  UpdateDBEntry = 'UpdateDBEntry',
+  // Start the save spinner
+  StartSave = 'StartSave',
+}
+
+// Action definitions
+type Action = (
+  | {
+    // Action type
+    type: ActionType.UpdateDBEntry,
+    // New state of the DBEntry
+    DBEntry: DBEntry,
+  }
+  | {
+    // Action type
+    type: (
+      | ActionType.StartSave
+    ),
+  }
+);
+
+/**
+ * Reducer that executes actions
+ * @author Yuen Ler Chow
+ * @param state current state
+ * @param action action to execute
+ */
+const reducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case ActionType.UpdateDBEntry: {
+      return {
+        ...state,
+        entry: action.DBEntry,
+      };
+    }
+    case ActionType.StartSave: {
+      return {
+        ...state,
+        saving: true,
+      };
+    }
+    default: {
+      return state;
+    }
+  }
+};
+
+/*------------------------------------------------------------------------*/
+/* ------------------------------ Component ----------------------------- */
+/*------------------------------------------------------------------------*/
+
+const AddorEditDBEntry: React.FC<Props> = (props) => {
+  /*------------------------------------------------------------------------*/
+  /* -------------------------------- Setup ------------------------------- */
+  /*------------------------------------------------------------------------*/
+
+  /* -------------- Props ------------- */
+
+  // Destructure all props
+  const {
+    onFinished,
+    entryFields,
+    DBEntryToEdit,
+    validationFunction,
+    objectModifier,
+    idPropName,
+    endpoint,
+    entries,
+  } = props;
+
+  /* -------------- State ------------- */
+
+  // Initial state
+  const initialState: State = {
+    entry: DBEntryToEdit || {},
+    saving: false,
+  };
+
+  // Initialize state
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  // Destructure common state
+  const {
+    entry,
+    saving,
+  } = state;
+
+  /*------------------------------------------------------------------------*/
+  /* ------------------------- Component Functions ------------------------ */
+  /*------------------------------------------------------------------------*/
+
+  /**
+   * Save changes to the DBEntry and then finish
+   * @author Yuen Ler Chow
+   */
+  const save = async () => {
+    // Start the save loading indicator
+    dispatch({ type: ActionType.StartSave });
+
+    // add all default values to the entry
+    entryFields.forEach((field) => {
+      if (field.defaultValue && !entry[field.objectKey]) {
+        entry[field.objectKey] = field.defaultValue;
+      }
+    });
+
+    const modifiedEntry = objectModifier ? objectModifier(entry) : entry;
+
+    // Send to server
+    try {
+      await visitServerEndpoint({
+        path: endpoint,
```

Comment:
Could we call endpoint something more descriptive? Maybe `saveEndpointPath` or something?